{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load annotation file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>original_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tig00000000</td>\n",
       "      <td>43327</td>\n",
       "      <td>82701</td>\n",
       "      <td>20281-Ct.00g000010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tig00000000</td>\n",
       "      <td>92730</td>\n",
       "      <td>150361</td>\n",
       "      <td>20281-Ct.00g000020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tig00000000</td>\n",
       "      <td>156651</td>\n",
       "      <td>158870</td>\n",
       "      <td>20281-Ct.00g000030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tig00000000</td>\n",
       "      <td>158921</td>\n",
       "      <td>159397</td>\n",
       "      <td>20281-Ct.00g000040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tig00000000</td>\n",
       "      <td>159552</td>\n",
       "      <td>160696</td>\n",
       "      <td>20281-Ct.00g000050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43900</th>\n",
       "      <td>tig00108041</td>\n",
       "      <td>23229</td>\n",
       "      <td>26330</td>\n",
       "      <td>20281-Ct.00g439060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43901</th>\n",
       "      <td>tig00108042</td>\n",
       "      <td>8670</td>\n",
       "      <td>9796</td>\n",
       "      <td>20281-Ct.00g439070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43902</th>\n",
       "      <td>tig00108042</td>\n",
       "      <td>14053</td>\n",
       "      <td>14259</td>\n",
       "      <td>20281-Ct.00g439080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43903</th>\n",
       "      <td>tig00108042</td>\n",
       "      <td>14297</td>\n",
       "      <td>19693</td>\n",
       "      <td>20281-Ct.00g439090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43904</th>\n",
       "      <td>tig00108042</td>\n",
       "      <td>24132</td>\n",
       "      <td>25730</td>\n",
       "      <td>20281-Ct.00g439100</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>43905 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene   start     end       original_name\n",
       "0      tig00000000   43327   82701  20281-Ct.00g000010\n",
       "1      tig00000000   92730  150361  20281-Ct.00g000020\n",
       "2      tig00000000  156651  158870  20281-Ct.00g000030\n",
       "3      tig00000000  158921  159397  20281-Ct.00g000040\n",
       "4      tig00000000  159552  160696  20281-Ct.00g000050\n",
       "...            ...     ...     ...                 ...\n",
       "43900  tig00108041   23229   26330  20281-Ct.00g439060\n",
       "43901  tig00108042    8670    9796  20281-Ct.00g439070\n",
       "43902  tig00108042   14053   14259  20281-Ct.00g439080\n",
       "43903  tig00108042   14297   19693  20281-Ct.00g439090\n",
       "43904  tig00108042   24132   25730  20281-Ct.00g439100\n",
       "\n",
       "[43905 rows x 4 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Load gff3\n",
    "use_cols1 = [0,3,4,8]\n",
    "df1 = pd.read_csv('ct_manu_gene.gff3',sep='\\t',header=None,usecols=use_cols1,encoding=\"utf-8-sig\")\n",
    "df1.columns = ['gene','start','end','info']\n",
    "df1['original_name'] = df1['info'].apply(lambda x: x.split(';')[-1].split('=')[-1])\n",
    "df1.drop('info',axis=1,inplace=True)\n",
    "df1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load vcf file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>pos</th>\n",
       "      <th>annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tig00000000</td>\n",
       "      <td>157455</td>\n",
       "      <td>ANN=A|missense_variant|MODERATE|Ct.00g000030|C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tig00000000</td>\n",
       "      <td>157456</td>\n",
       "      <td>ANN=A|missense_variant|MODERATE|Ct.00g000030|C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tig00000000</td>\n",
       "      <td>157459</td>\n",
       "      <td>ANN=A|missense_variant|MODERATE|Ct.00g000030|C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tig00000000</td>\n",
       "      <td>157462</td>\n",
       "      <td>ANN=A|missense_variant|MODERATE|Ct.00g000030|C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tig00000000</td>\n",
       "      <td>157465</td>\n",
       "      <td>ANN=T|missense_variant|MODERATE|Ct.00g000030|C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>457382</th>\n",
       "      <td>tig00108041</td>\n",
       "      <td>23309</td>\n",
       "      <td>ANN=C|missense_variant|MODERATE|Ct.00g439010|C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>457383</th>\n",
       "      <td>tig00108041</td>\n",
       "      <td>23310</td>\n",
       "      <td>ANN=G|synonymous_variant|LOW|Ct.00g439010|Ct.0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>457384</th>\n",
       "      <td>tig00108041</td>\n",
       "      <td>23313</td>\n",
       "      <td>ANN=A|synonymous_variant|LOW|Ct.00g439010|Ct.0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>457385</th>\n",
       "      <td>tig00108041</td>\n",
       "      <td>23316</td>\n",
       "      <td>ANN=A|missense_variant|MODERATE|Ct.00g439010|C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>457386</th>\n",
       "      <td>tig00108041</td>\n",
       "      <td>23319</td>\n",
       "      <td>ANN=T|synonymous_variant|LOW|Ct.00g439010|Ct.0...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>457387 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               gene     pos                                         annotation\n",
       "0       tig00000000  157455  ANN=A|missense_variant|MODERATE|Ct.00g000030|C...\n",
       "1       tig00000000  157456  ANN=A|missense_variant|MODERATE|Ct.00g000030|C...\n",
       "2       tig00000000  157459  ANN=A|missense_variant|MODERATE|Ct.00g000030|C...\n",
       "3       tig00000000  157462  ANN=A|missense_variant|MODERATE|Ct.00g000030|C...\n",
       "4       tig00000000  157465  ANN=T|missense_variant|MODERATE|Ct.00g000030|C...\n",
       "...             ...     ...                                                ...\n",
       "457382  tig00108041   23309  ANN=C|missense_variant|MODERATE|Ct.00g439010|C...\n",
       "457383  tig00108041   23310  ANN=G|synonymous_variant|LOW|Ct.00g439010|Ct.0...\n",
       "457384  tig00108041   23313  ANN=A|synonymous_variant|LOW|Ct.00g439010|Ct.0...\n",
       "457385  tig00108041   23316  ANN=A|missense_variant|MODERATE|Ct.00g439010|C...\n",
       "457386  tig00108041   23319  ANN=T|synonymous_variant|LOW|Ct.00g439010|Ct.0...\n",
       "\n",
       "[457387 rows x 3 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Load vcf\n",
    "use_cols2 = [0,1,7]\n",
    "df2 = pd.read_csv('ct_ann.vcf',delimiter='\\t',header=None,usecols=use_cols2,encoding=\"utf-8-sig\")\n",
    "df2.columns = ['gene','pos','annotation']\n",
    "df2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Getting an intermediate file with SNPs in the genic regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load significant gene file\n",
    "#use_cols = [0]  #this part depends on the format of input sig files\n",
    "\n",
    "#df3 = pd.read_csv('Significant_CULTRF_2000static_vars_PhenoCol-08.assoc.txt_NamesOnly.txt',header=None,usecols=use_cols,sep=' ')\n",
    "df3 = pd.read_csv('Significant_soil_rooting_conditionsstatic_vars_PhenoCol-38.assoc.txt_NamesOnly.txt',sep='\\t')\n",
    "df3 = pd.DataFrame(df3.index)\n",
    "df3.columns = ['Gene']\n",
    "\n",
    "#df3 = pd.read_csv('ctar_MODCF-cloud-cov_soft2FDR_2020-06-11.txt',delimiter='\\t',encoding=\"utf-8-sig\")\n",
    "#print(df3.columns.tolist())\n",
    "uniq_max = df3.Gene.unique()\n",
    "\n",
    "## getting variants in gene regions\n",
    "\n",
    "d = {} # dictionary to hold intersected genes with the position of SNPs\n",
    "l = [] # list to hold significant SNPs found in the intersection of gff and vcf\n",
    "\n",
    "for item in uniq_max:  # iterating through the unique elements in the significant gene list\n",
    "    gene1 = item.split(':')[0]      # getting contig name\n",
    "    pos1 = item.split(':')[1]\n",
    "    \n",
    "    if gene1 in list(df1['gene']):  #check if the contig name matches in gff\n",
    "        df5 = df1[df1['gene']==gene1]  # create a new dataframe of gff with the selected gene \n",
    "        df5.reset_index(drop=True, inplace=True)  #drop old index. VERY IMPORTANT!!!\n",
    "    \n",
    "    else:\n",
    "        continue\n",
    "    \n",
    "    for i in range(len(df5)):\n",
    "        if df5['start'][i]-100 <= int(pos1) <= df5['end'][i]+100: #check if correlated position falls inside a feature\n",
    "            if item not in l:  #ignoring already added genes\n",
    "                l.append(item)\n",
    "            gene2 = df5['gene'][i] + ':' + str(df5['start'][i]) + ':' + str(df5['end'][i])+'+'+df5['original_name'][i] #gene found \n",
    "            if gene2 not in d.keys():\n",
    "                d[gene2] = [pos1]\n",
    "            else:\n",
    "                d[gene2].append(pos1)\n",
    "        \n",
    "\n",
    "sig_list = [] # hold matched significant SNPs\n",
    "gen_list = [] # hold intersected genes\n",
    "snp_list = [] # hold number of SNPs found within intersected gene\n",
    "\n",
    "for i in l: \n",
    "    for j in list(d.keys()):\n",
    "        gen1 = i.split(':')\n",
    "        gen2 = j.split('+')[0].split(':')\n",
    "        if gen1[0] == gen2[0] and int(gen2[1])-100 <= int(gen1[1]) <= int(gen2[2])+100:\n",
    "            sig_list.append(i)\n",
    "            gen_list.append(j)\n",
    "            snp_list.append(len(d[j]))\n",
    "            \n",
    "elem = [sig_list,gen_list,snp_list]\n",
    "df_temp_min = pd.DataFrame(elem).T\n",
    "df_temp_min.columns = ['Significant SNPs','Genes','Number of SNPs in the gene']\n",
    "df_temp_min['Genes with start-end positions'] = df_temp_min['Genes'].apply(lambda x: x.split('+')[0])\n",
    "df_temp_min['Original_name'] = df_temp_min['Genes'].apply(lambda x: x.split('+')[1])\n",
    "df_temp_min_copy = df_temp_min.copy()\n",
    "df_temp_min_copy.drop('Genes',axis=1,inplace=True)\n",
    "df_temp_min_copy = df_temp_min_copy[['Significant SNPs','Genes with start-end positions','Original_name','Number of SNPs in the gene']]\n",
    "df_temp_min_copy.to_csv('aet.txt',sep='\\t')  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Getiing gene function from InterproScan file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tig00000000</td>\n",
       "      <td>43327</td>\n",
       "      <td>43433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tig00000000</td>\n",
       "      <td>43488</td>\n",
       "      <td>43754</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tig00000000</td>\n",
       "      <td>58736</td>\n",
       "      <td>58866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tig00000000</td>\n",
       "      <td>58898</td>\n",
       "      <td>59237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tig00000000</td>\n",
       "      <td>65151</td>\n",
       "      <td>65280</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             0      3      4\n",
       "0  tig00000000  43327  43433\n",
       "1  tig00000000  43488  43754\n",
       "2  tig00000000  58736  58866\n",
       "3  tig00000000  58898  59237\n",
       "4  tig00000000  65151  65280"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Name    IPR Term        IPR Description Source  Source Term     Source Description      GO\n",
    "use_cols1 = ['Name','Source Description']\n",
    "\n",
    "dff3 = pd.read_csv('interpro.tab',sep='\\t',usecols=use_cols1)\n",
    "dff3['new_name'] = dff3['Name'].apply(lambda x: x[6:-4])\n",
    "dff3.dropna(inplace=True)\n",
    "dff3.reset_index(drop=True, inplace=True)\n",
    "\n",
    "use_cols11 = [0,3,4]\n",
    "dfexon = pd.read_csv('exxon.gff3',sep='\\t',header=None,usecols=use_cols11,encoding=\"utf-8-sig\")\n",
    "dfexon.head()\n",
    "\n",
    "#dff3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "func = []\n",
    "gene = []\n",
    "di = {}\n",
    "ds = {}\n",
    "dic = {}\n",
    "\n",
    "dff1 = pd.read_csv('aet.txt',sep='\\t',index_col=0)\n",
    "\n",
    "for gene1 in dff1['Original_name']:\n",
    "    dff4 = dff3[dff3['new_name'] == gene1[6:]]\n",
    "    dff4.reset_index(drop=True, inplace=True)  #drop old index. VERY IMPORTANT!!! \n",
    "    for i in range(len(dff4)):\n",
    "        #snp = dff1['Significant SNPs'][i]\n",
    "        if dff4['new_name'][i] == gene1[6:]:\n",
    "            if gene1 not in di:\n",
    "                if dff4['Source Description'][i] not in ['FAMILY NOT NAMED','SUBFAMILY NOT NAMED']:\n",
    "                    di[gene1] = [dff4['Source Description'][i]]\n",
    "            \n",
    "            else:\n",
    "                if dff4['Source Description'][i] not in di[gene1]:\n",
    "                    if dff4['Source Description'][i] not in ['FAMILY NOT NAMED','SUBFAMILY NOT NAMED']:\n",
    "                        di[gene1].append(dff4['Source Description'][i])\n",
    "                else:\n",
    "                    continue\n",
    "                    \n",
    "for k,v in di.items():\n",
    "    dic[k] = [' / '.join(di[k])]\n",
    "dic\n",
    "\n",
    "entry = dic.keys()\n",
    "entry\n",
    "\n",
    "dfw = pd.DataFrame(dic).T\n",
    "\n",
    "fun = []\n",
    "for gene in df_temp_min_copy['Original_name']:\n",
    "    if gene not in dfw.index:\n",
    "        fun.append('Gene not found in interproscan')\n",
    "        continue\n",
    "    for index in dfw.index:\n",
    "        if gene == index:\n",
    "            fun.append(dfw[0][index])\n",
    "            \n",
    "\n",
    "df_temp_min_copy['Gene Function'] = fun\n",
    "df_temp_min_copy = df_temp_min_copy[['Significant SNPs','Genes with start-end positions','Original_name','Gene Function','Number of SNPs in the gene']]\n",
    "df_temp_min_copy.to_csv('result_wind.csv',header=True,sep=',',index=None)\n",
    "\n",
    "df_ = pd.read_csv('result_wind.csv')\n",
    "#df_.head()\n",
    "df_['Annotation'] = 'not_in_exon'\n",
    "df_['Amino acid change'] = np.nan\n",
    "df_['Nucleotide change'] = np.nan\n",
    "for i in range(len(df_)):\n",
    "    contig = df_['Significant SNPs'][i].split(':')\n",
    "    \n",
    "    df6 = df2[df2['gene']==contig[0]]  # create a new dataframe of gff with the selected gene \n",
    "    df7 = dfexon[dfexon[0]==contig[0]]\n",
    "    df6.reset_index(drop=True, inplace=True)\n",
    "    df7.reset_index(drop=True, inplace=True)\n",
    "    \n",
    "    for j in range(len(df7)):\n",
    "        if int(df7[3][j]) <= int(contig[1]) <= int(df7[4][j]):\n",
    "            for k in range(len(df6)):\n",
    "                if int(df6['pos'][k]) == int(contig[1]):\n",
    "                    df_.loc[i,'Annotation'] = df6['annotation'][k].split('|')[1]\n",
    "                    df_.loc[i,'Amino acid change'] = df6['annotation'][k].split('|')[10]\n",
    "                    df_.loc[i,'Nucleotide change'] = df6['annotation'][k].split('|')[9]\n",
    "        \n",
    "df_.to_csv('result_soil_rooting_conditions.csv',header=True,sep=',',index=None)   \n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
